mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -830,17 +830,617 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
+    }
+    TransformMapAttr twoDimLDSView = tensorToLDSViewBuilder.get();
+
+    TopDownTMBuilder flatLDSViewBuilder =
+        TopDownTMBuilder::below(tensorToLDSViewBuilder, twoDimLDSView);
+    flatLDSViewBuilder.unmerge(
+        "flatDim", 0, {"nrDim", "rDim"},
+        {nonReduceMergeDimSize, lowestShape[reduceAxis]});
+    TransformMapAttr flatLDSView = flatLDSViewBuilder.get();
+    SmallVector<Attribute> threadsToLDSViewAttrs;
+    for (Attribute trMap : regTensorView) {
+      threadsToLDSViewAttrs.push_back(trMap);
+    }
+    threadsToLDSViewAttrs.push_back(twoDimLDSView);
+    threadsToLDSViewAttrs.push_back(flatLDSView);
+    return rewriter.getArrayAttr(threadsToLDSViewAttrs);
+  }
+
+  // This should only be used if product non-reduction dims is
+  // equal or larger than number threads in a block.
+  //
+  // Given a input tensor : D0, ... , Dr , ... , DN to reduce,
+  // This function creates a view that maps the space of
+  // [D0, ... , Dr , ... , DN] --> [tid, nrIter, rIter] where
+  // tid is threads within the block, nrIter is non-reducing
+  // iterations within a thread and rIter is reducing iterations
+  // within a thread.
+  ArrayAttr createThreadViewForNRLargerThanThreads(
+      Location loc, ArrayRef<int64_t> toReduceShape, int64_t blockSize,
+      int64_t reduceAxis, PatternRewriter &rewriter) const {
+    BottomUpTMBuilder threadsToTensor(rewriter, toReduceShape, loc);
+    SmallVector<StringRef, 4> lowerNameRefs;
+    threadsToTensor.getStartNames(lowerNameRefs);
+
+    int64_t nonReduceMergeDimSize = 1;
+    SmallVector<StringRef, 4> nonReduceNameRefs;
+    for (auto dimAndSize : llvm::enumerate(toReduceShape)) {
+      int64_t dim = dimAndSize.index();
+      int64_t dimSize = dimAndSize.value();
+      if (dim != reduceAxis) {
+        nonReduceMergeDimSize *= dimSize;
+        nonReduceNameRefs.push_back(lowerNameRefs[dim]);
+      }
+    }
+    threadsToTensor.merge("nrDim", 0, nonReduceNameRefs);
+    threadsToTensor.passThrough({"rIter"}, {1}, {lowerNameRefs[reduceAxis]});
+    TransformMapAttr mergeTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, mergeTrMap);
+    int64_t nrThreads = (nonReduceMergeDimSize + (blockSize - 1)) / blockSize;
+    threadsToTensor.pad({"nrDim"},
+                        {0, blockSize * nrThreads - nonReduceMergeDimSize});
+    threadsToTensor.passThrough({"rIter"}, {1}, {"rIter"});
+    TransformMapAttr padTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, padTrMap);
+    threadsToTensor.unmerge({"tid", "nrIter"}, {0, 1}, "nrDim",
+                            {blockSize, nrThreads});
+    threadsToTensor.passThrough({"rIter"}, {2}, {"rIter"});
+    TransformMapAttr unmergeTrMap = threadsToTensor.get();
+
+    return rewriter.getArrayAttr({unmergeTrMap, padTrMap, mergeTrMap});
+  }
+
+  // This should only be used if product non-reduction dims is
```

Comment:
Can this and the function above be combined in some reasonable way?

Reply:
It does two different things even from the first transform?